docs/rfc/rfc-002-ipc-ecosystem.md

```diff
@@ -0,0 +1,348 @@
+code and the developer experience, and hides bugs. To support a more robust,
+trustworthy, and usable system, we should document which communication paths
+are essential, which could be removed or reduced in scope, and what we can
+improve for the most important use cases.
+
+
+## Background
+
+The Tendermint state replication engine has a complex IPC footprint.
+
+1. Consensus nodes communicate with each other using a networked peer-to-peer
+   message-passing protocol.
+
+2. Consensus nodes communicate with the application whose state is being
+   replicated via the [Application BlockChain Interface (ABCI)][abci].
+
+3. Consensus nodes export a network-accessible [RPC service][rpc-service] to
+   support operations (bootstrapping, debugging) and synchronization of [light clients][light-client].
+   This interface is also used by the [`tendermint` CLI][tm-cli].
+
+4. Consensus nodes export a gRPC service exposing a subset of the methods of
+   the RPC service described by (3). This was intended to simplify the
+   implementation of tools that already use gRPC to communicate with an
+   application (via the Cosmos SDK), and wanted to also talk to the consensus
+   node without implementing yet another RPC protocol.
+
+   The gRPC interface to the consensus node has been deprecated and is slated
+   for removal in the forthcoming Tendermint v0.36 release.
+
+5. Consensus nodes may optionally communicate with a "remote signer" that holds
+   a validator key and can provide public keys and signatures to the consensus
+   node. One of the stated goals of this configuration is to allow the signer
+   to be run on a private network, separate from the consensus node, so that a
+   compromise of the consensus node from the public network would be less
+   likely to expose validator keys.
+
+## Discussion: Transport Mechanisms
+
+### Remote Signer Transport
+
+A remote signer communicates with the consensus node in one of two ways:
+
+1. "Raw": Using a TCP or Unix-domain socket which carries varint-prefixed
+   protocol buffer messages. In this mode, the consensus node is the server,
+   and the remote signer is the client.
+
+   This mode has been deprecated, and is intended to be removed.
+
+2. gRPC: This mode uses the same protobuf messages as "Raw" node, but uses a
+   standard encrypted gRPC HTTP/2 stub as the transport. In this mode, the
+   remote signer is the server and the consensus node is the client.
+
+TODO: It's not clear to me how meaningfully a remote signer is really isolated
+by this mechanism. The consensus node has to be on a shared network with the
+remote signer, so a compromise of the consensus node from the public network
+leaves the attacker able to communicate with the remote signer anyway. It does
+raise the bar by a constant factor -- now you also have to compromise the
+remote signer too -- but it's not actually isolated in the gRPC configuration.
```

Comment:
But what can an attacker do that compromises the nodes do with the remote signer. The logic within the remote signer will prevent any form of double signing. The attacker can mimic the node (i.e. generate arbitrary signed messages) but still can't access the private keys. Also another function of a remote signer is to support a redundant validator strategy whereby multiple validators are running and if one fails the next one can fill in using the same remote signer.

Reply:
> But what can an attacker do that compromises the nodes do with the remote signer. 

If the attacker can reach the remote signer, it also becomes possible to attack and compromise the signer process too, since the signer is no longer network-isolated from the attacker once they control the consensus node. The indirection certainly slows the attacker down (like a second lock on the front door), but it doesn't prevent a compromise of key material. (To be clear, I'm _not_ advocating we get rid of this)

Reply:
I'm going to remove this TODO. I wrote it to jot down my thoughts as I was reading the remote signer docs and code, but it's not really germane to the topic of this RFC.